Adds overloads that use TimeSpan instead of Int32.

MicahZoltu opened the pull request:
I tried to follow the conventions elsewhere in the file, but feel free to let me know if I have not followed code style correctly.

The `ArgumentOutOfRangeException` seemed like the best way to handle the fact that `TimeSpan`s aren't really bounded, though with the new the API surface area the upper bound constraint seems arbitrary.  However, in most real world use cases it will probably not be hit and it is documented.  I decided to throw on negative number as well because that doesn't make sense.  The overload doesn't accept null because if you want to pass null explicitly you can simply call the other overload (transparent to the user) and if you have a `TimeSpan?` you are doing it wrong.  ;)

I added two tests, but I'll admit that they are pretty weak.  I didn't see any tests for the original overloads and I'm not sure how to author a test that will correctly spin up a `DatabaseFacade` where the `SetCommandTimeout` will actually succeed.  Feedback is welcome on how I can make a positive test for timeout setting.

The overload that returns a `TimeSpan` is likely a bit contentious, as is always the case when you have an "overload" that differs only by return type.  I can pull it out if desired, I added it mainly for consistency but I admit that it feels quite awkward.

Fixes #6614

smitpatel commented on src/Microsoft.EntityFrameworkCore.Relational/RelationalDatabaseFacadeExtensions.cs:
```diff
@@ -211,9 +211,29 @@ public static IDbContextTransaction BeginTransaction([NotNull] this DatabaseFaca
+            {
+                throw new ArgumentOutOfRangeException($"Timeout must be less than or equal to Int32.MaxValue (2147483647) milliseconds.  Provided: {timeout.TotalMilliseconds} milliseconds.");
+            }
+            SetCommandTimeout(databaseFacade, Convert.ToInt32(timeout.TotalMilliseconds));
```
Should there be factor of 1000 somewhere since timeout in SqlConnection is in seconds rather than milliseconds.

MicahZoltu replied:
If it is in Seconds I'll change it to `TotalSeconds`.  The whole reason I am creating this PR is because I have no idea what units `timeout` is actually in.  😃

smitpatel replied:
👍 
Here is more info on ConnectionTimeout in SqlConnection.
https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlconnection.connectiontimeout(v=vs.110).aspx

MicahZoltu commented:
I recently changed my GitHub username from `Zoltu` to `MicahZoltu` (I created a new GitHub Organization named `Zoltu`).  I have previously done the whole CLA thing, but it appears that the bot didn't follow my name change.  What is the correct way to go about resolving that?

Side-note: The bot should really associate with users by ID, not by name.

dnfclas commented:
__@MicahZoltu__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

smitpatel commented:
@MicahZoltu - Tests for original overload are here
https://github.com/aspnet/EntityFramework/blob/c9b5916a4e0fff763febea00da793ebd4113632d/test/Microsoft.EntityFrameworkCore.SqlServer.Tests/CommandConfigurationTests.cs

You can add tests in the same file which uses different way to configure them and verify if they are set correctly.

MicahZoltu commented:
I moved the tests into `CommandConfigurationTests.cs` (additional asserts in existing tests).  I also changed the exception thrown to `ArgumentException` to match the existing exception type.

I also made the return value of `GetCommandTimeoutTimeSpan` nullable to match the existing behavior (and address a bug in null handling).

smitpatel commented on src/Microsoft.EntityFrameworkCore.Relational/RelationalDatabaseFacadeExtensions.cs:
```diff
@@ -211,9 +211,36 @@ public static IDbContextTransaction BeginTransaction([NotNull] this DatabaseFaca
+            }
+            if (timeout.TotalSeconds > Int32.MaxValue)
+            {
+                throw new ArgumentException($"Timeout must be less than or equal to Int32.MaxValue (2147483647) milliseconds.  Provided: {timeout.Seconds} seconds.");
```
not `milliseconds` in message.

smitpatel commented:
@rowanmiller - Can you scan this for API review? Milestone?

MicahZoltu commented on src/Microsoft.EntityFrameworkCore.Relational/RelationalDatabaseFacadeExtensions.cs:
```diff
@@ -211,9 +211,36 @@ public static IDbContextTransaction BeginTransaction([NotNull] this DatabaseFaca
+        /// <param name="databaseFacade">The database facade to set the timeout on.</param>
+        /// <param name="timeout">The time to wait before the command is aborted.</param>
+        /// <exception cref="ArgumentOutOfRangeException">If the provided timeout is greater than Int32.MaxValue (seconds) or less than 0.</exception>
+        public static void SetCommandTimeout([NotNull] this DatabaseFacade databaseFacade, TimeSpan timeout)
```
Fixed.

smitpatel commented on src/Microsoft.EntityFrameworkCore.Relational/RelationalDatabaseFacadeExtensions.cs:
```diff
@@ -211,9 +211,36 @@ public static IDbContextTransaction BeginTransaction([NotNull] this DatabaseFaca
 
         public static int? GetCommandTimeout([NotNull] this DatabaseFacade databaseFacade)
             => GetRelationalConnection(databaseFacade).CommandTimeout;
+        public static TimeSpan? GetCommandTimeoutTimeSpan([NotNull] this DatabaseFacade databaseFacade)
```
@rowanmiller - Did we take any decision on this?

MicahZoltu replied:
The only reason I can think of where I would want to get this value (rather than just set it) is if I were building a generalized monitoring library.  Since my main concern is getting the `TimeSpan` setter, I'm willing to ignore the monitoring story in an effort to move the rest of this PR forward.

MicahZoltu replied:
Removed.  However, I did it from the GitHub GUI since I'm not at my home computer, if the build doesn't pass I may have to wait until this evening to test it locally.

rowanmiller commented:
@smitpatel this is good to go from my perspective.

@MicahZoltu sorry that I held this up, it had dropped of my radar of things to do.

smitpatel commented:
@MicahZoltu - Can you squash all commits and rebase on dev?

MicahZoltu commented:
Sure, though it'll have to wait until this evening when I'm at my computer.

MicahZoltu commented:
Rebased and squashed.

rowanmiller commented:
@smitpatel can you get this one merged now

smitpatel commented:
Merged in 3608b72f59ff110d75672569ea84e82288de6ab8
Thanks for contribution @MicahZoltu
